### PR TITLE
[Feat] better kornia.enhance testing

### PR DESCRIPTION
**Pull request by edgarriba**
### Description

Improves the testing for koria.enhance for tpu+jit

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/utils/image.py [hunk 1]
```diff
@@ -69,10 +69,12 @@ def _to_bchw(tensor: torch.Tensor, color_channel_num: Optional[int] = None) -> t
...
-    if color_channel_num is not None and color_channel_num != 1:
-        channel_list = [0, 1, 2, 3]
-        channel_list.insert(1, channel_list.pop(color_channel_num))
-        tensor = tensor.permute(*channel_list)
+    # TODO(jian): this function is never used. Besides is not feasible for torchscript.
+    # In addition, the docs must be updated. I don't understand what is doing.
+    # if color_channel_num is not None and color_channel_num != 1:
+    #    channel_list = [0, 1, 2, 3]
+    #    channel_list.insert(1, channel_list.pop(color_channel_num))
+    #    tensor = tensor.permute(*channel_list)
```
Comment (edgarriba) [inline kornia/utils/image.py:77]: @shijianjian could you verify this ?

**Comment by edgarriba** [comment]
@shijianjian check the logs - in the function `adjust_hue_raw` there's a checking that seems that xla device don't like. I think we should we remove the bound checks inside tensors.

File: kornia/utils/image.py (hunk 1 above)
Comment (shijianjian) [reply kornia/utils/image.py:77]: This is a more general function that in case the input is (B, H, W, C) or other weird formats.

So, if (B, H, W, C), we will need to have "color_channel_num == 3" to convert it (B, C, H, W).

**Comment by shijianjian** [comment]
I could fix the hue error, it is only a tensor conversion error. But not sure what to do for ZCA.

File: kornia/utils/image.py (hunk 1 above)
Comment (edgarriba) [reply kornia/utils/image.py:77]: Is it really used ? And really needed ? Couldn't raise any breakpoint in tests to verif.  Besides, I think the documentation between 2d and 3d have typo.

**Comment by edgarriba** [comment]
@shijianjian I'll decrease the precision tolerance but in general I see that xla produce less precise results compared to CPUs or GPUs

**Comment by edgarriba** [comment]
@shijianjian could you review ? Mostly increased test coverage and update docs. I added few todos since some functionalities are not currently working with xla devices.

File: kornia/enhance/normalize.py [hunk 2]
```diff
@@ -57,32 +63,47 @@ def __repr__(self):
...
-    Args:
-        data (torch.Tensor): The image tensor to be normalised.
-        mean (torch.Tensor or float): Mean for each channel.
-        std (torch.Tensor or float): Standard deviations for each channel.
+    .. math::
+        \text{input[channel] = (input[channel] - mean[channel]) / std[channel]}
 
-    Returns:
-        torch.Tensor: The normalised image tensor.
+    Where `mean` is :math:`(M_1,...,M_n)` and `std` :math:`(S_1,..,S_n)` for `n` channels,
```
Comment (shijianjian) [inline kornia/enhance/normalize.py:71]: Anything missing after the comma?

File: test/enhance/test_adjust.py [hunk 3]
```diff
@@ -42,38 +43,37 @@ def test_saturation_one_batch(self, device, dtype):
...
-        img = torch.rand(batch_size, channels, height, width)
+        img = torch.rand(batch_size, channels, height, width, device=device, dtype=dtype)
         img = tensor_to_gradcheck_var(img)  # to var
         assert gradcheck(kornia.adjust_saturation, (img, 2.),
                          raise_exception=True)
 
 
 class TestAdjustHue:
-    def test_hue_one(self, device):
+    def test_hue_one(self, device, dtype):
```
Comment (shijianjian) [inline test/enhance/test_adjust.py:61]: Just wandering if to have a BaseTester for all modulized functions to perform some routined tests as well. /cc @FGeri

**Review by shijianjian — APPROVED** [review]
LGTM. Just wandering what tests shall be included for all the nn.modules. Like ONNX export and load?

File: kornia/enhance/adjust.py [hunk 4]
```diff
@@ -36,20 +37,21 @@ def adjust_saturation_raw(input: torch.Tensor, saturation_factor: Union[float, t
...
+    if not isinstance(input, torch.Tensor):
         raise TypeError(f"Input type is not a torch.Tensor. Got {type(input)}")
 
     if not isinstance(saturation_factor, (float, torch.Tensor,)):
         raise TypeError(f"The saturation_factor should be a float number or torch.Tensor."
                         f"Got {type(saturation_factor)}")
 
     if isinstance(saturation_factor, float):
-        saturation_factor = torch.tensor([saturation_factor])
+        saturation_factor = torch.as_tensor(saturation_factor)
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/enhance/adjust.py:48]: this happens in line 50. It also forces that the saturation factor goes to same device and dtype as the input in case it comes as tensor. Maybe we can simplify this cases as you suggest and in case is a tensor with a different device or dtype as input, let pytorch raise an error. Might be easier later to detect potential bugs.

File: kornia/enhance/adjust.py [hunk 5]
```diff
@@ -36,20 +37,21 @@ def adjust_saturation_raw(input: torch.Tensor, saturation_factor: Union[float, t
...
 
     saturation_factor = saturation_factor.to(input.device).to(input.dtype)
 
-    if (saturation_factor < 0).any():
-        raise ValueError(f"Saturation factor must be non-negative. Got {saturation_factor}")
+    # TODO: find a proper way to check bound values in batched tensors.
+    # if (saturation_factor < 0).any():
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/enhance/adjust.py:53]: xla was complaining, and I believe for large tensor with operation could harm a bit the performance. I'm not quite sure that we have to do bound checking inside tensors.

File: test/enhance/test_adjust.py (hunk 3 above)
Comment (edgarriba) [reply test/enhance/test_adjust.py:61]: we can have a seconf iteration for BaseTester and automate a lot of things
